File: Source/webapp/src/lib/image-quality.test.ts
import { describe, it, expect } from 'vitest'
import { assessImageQuality, qualityIssueMessage, type LuminanceSource } from './image-quality'

// Build an RGBA buffer from a per-pixel gray value function (r=g=b=gray, a=255).
function grayImage(width: number, height: number, gray: (x: number, y: number) => number): LuminanceSource {
  const data = new Uint8ClampedArray(width * height * 4)
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const g = Math.max(0, Math.min(255, Math.round(gray(x, y))))
      const i = (y * width + x) * 4
      data[i] = g
      data[i + 1] = g
      data[i + 2] = g
      data[i + 3] = 255
    }
  }
  return { data, width, height }
}

// A sharp, well-exposed image: a high-contrast checkerboard around mid-grey.
const sharp = grayImage(32, 32, (x, y) => ((x + y) % 2 === 0 ? 40 : 210))

describe('assessImageQuality', () => {
  it('passes a sharp, well-exposed image', () => {
    expect(assessImageQuality(sharp)).toEqual({ ok: true, issues: [] })
  })

  it('flags a flat (blurred) image as BLURRY', () => {
    const flat = grayImage(32, 32, () => 128) // no edges → ~zero Laplacian variance
    const v = assessImageQuality(flat)
    expect(v.ok).toBe(false)
    expect(v.issues).toContain('BLURRY')
  })

  it('flags a washed-out image as GLARE', () => {
    // Mostly blown-out white with a little texture to keep it sharp.
    const glary = grayImage(32, 32, (x, y) => ((x + y) % 2 === 0 ? 250 : 255))
    const v = assessImageQuality(glary)
    expect(v.issues).toContain('GLARE')
  })

  it('flags a dark image as TOO_DARK', () => {
    const dark = grayImage(32, 32, (x, y) => ((x + y) % 2 === 0 ? 5 : 30)) // sharp but mean ~17
    const v = assessImageQuality(dark)
    expect(v.issues).toContain('TOO_DARK')
  })

  it('does not judge an empty image', () => {
    expect(assessImageQuality({ data: new Uint8ClampedArray(0), width: 0, height: 0 })).toEqual({ ok: true, issues: [] })
  })

  it('respects custom thresholds', () => {
    const flat = grayImage(32, 32, () => 128)
    // A permissive sharpness floor of 0 accepts the flat image.
    expect(assessImageQuality(flat, { minSharpness: 0, maxBrightFraction: 1, minMeanLuminance: 0 }).ok).toBe(true)
  })
})

describe('qualityIssueMessage', () => {
  it('is empty when there are no issues', () => {
    expect(qualityIssueMessage([])).toBe('')
  })

  it('names each issue and offers retake-or-override', () => {
    const msg = qualityIssueMessage(['BLURRY', 'GLARE'])
    expect(msg).toMatch(/blurry/)
    expect(msg).toMatch(/glare/)
    expect(msg).toMatch(/upload it anyway/)
  })
})

File: Source/mobile/test/image_quality_test.dart
import 'dart:math';
import 'dart:typed_data';

import 'package:flutter_test/flutter_test.dart';
import 'package:wobblio/core/domain/image_quality.dart';

// Build an RGBA buffer from a per-pixel grey value function (grey → R=G=B, A=255).
LuminanceSource _grey(int width, int height, int Function(int x, int y) value) {
  final data = Uint8List(width * height * 4);
  for (var y = 0; y < height; y++) {
    for (var x = 0; x < width; x++) {
      final i = (y * width + x) * 4;
      final v = value(x, y).clamp(0, 255);
      data[i] = v;
      data[i + 1] = v;
      data[i + 2] = v;
      data[i + 3] = 255;
    }
  }
  return LuminanceSource(data: data, width: width, height: height);
}

void main() {
  group('assessImageQuality', () {
    test('passes a sharp, well-lit, high-contrast image', () {
      // A checkerboard: strong edges (high Laplacian variance), mid-brightness, no glare.
      final src = _grey(32, 32, (x, y) => (x + y).isEven ? 40 : 200);
      final verdict = assessImageQuality(src);
      expect(verdict.ok, isTrue);
      expect(verdict.issues, isEmpty);
    });

    test('flags a flat (blur-equivalent) image as BLURRY', () {
      // A uniform mid-grey field has zero edge energy → Laplacian variance 0.
      final src = _grey(32, 32, (_, __) => 128);
      final verdict = assessImageQuality(src);
      expect(verdict.issues, contains(QualityIssue.blurry));
    });

    test('flags a washed-out image as GLARE', () {
      // Mostly blown-out (250) with a little edge texture so it is not also "blurry".
      final src = _grey(32, 32, (x, y) => (x + y).isEven ? 250 : 255);
      final verdict = assessImageQuality(src);
      expect(verdict.issues, contains(QualityIssue.glare));
    });

    test('flags a dark image as TOO_DARK', () {
      final rnd = Random(1);
      // Dark but textured (mean well below the 42 floor) so darkness — not blur — is the flag.
      final src = _grey(32, 32, (_, __) => rnd.nextInt(20));
      final verdict = assessImageQuality(src);
      expect(verdict.issues, contains(QualityIssue.tooDark));
    });

    test('does not judge a too-small image (fails open)', () {
      final src = _grey(2, 2, (_, __) => 128);
      expect(assessImageQuality(src).ok, isTrue);
    });
  });

  group('qualityIssueMessage', () {
    test('is empty for no issues', () {
      expect(qualityIssueMessage(const []), isEmpty);
    });

    test('renders an actionable retake tip that nudges laying the receipt flat', () {
      final msg = qualityIssueMessage(const [QualityIssue.blurry]);
      expect(msg, contains('blurry'));
      expect(msg, contains('Lay the receipt flat'));
      expect(msg, contains('upload it anyway'));
    });
  });
}
